server/redeemer.go

```diff
@@ -0,0 +1,303 @@
+	}, nil
+}
+
+func (r *redeemer) Start() error {
+	listener, err := gonet.Listen("tcp", ":50051")
+	defer listener.Close()
+	if err != nil {
+		return err
+	}
+	// slice of gRPC options
+	// Here we can configure things like TLS
+	opts := []grpc.ServerOption{}
+	// var s *grpc.Server
+	s := grpc.NewServer(opts...)
+	defer s.Stop()
+
+	net.RegisterTicketRedeemerServer(s, r)
+
+	return s.Serve(listener)
+}
+
+func (r *redeemer) Stop() {
+	close(r.quit)
+}
+
+func (r *redeemer) QueueTicket(ctx context.Context, ticket *net.Ticket) (*empty.Empty, error) {
+	t := pmTicket(ticket)
+	if err := r.sm.QueueTicket(t); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	go r.monitorMaxFloat(ethcommon.BytesToAddress(ticket.Sender))
+	return &empty.Empty{}, nil
+}
+
+func (r *redeemer) monitorMaxFloat(sender ethcommon.Address) {
+	_, ok := r.liveSenders.Load(sender)
+	if ok {
+		return
+	}
+	r.liveSenders.Store(sender, true)
+	sink := make(chan *big.Int, 10)
+	sub := r.sm.MonitorMaxFloat(sender, sink)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case <-r.quit:
+			return
+		case err := <-sub.Err():
+			glog.Error(err)
+		case mf := <-sink:
+			r.sendMaxFloatUpdate(sender, mf)
+		}
+	}
+}
+
+func (r *redeemer) sendMaxFloatUpdate(sender ethcommon.Address, maxFloat *big.Int) {
+	r.subs.Range(
+		func(key, value interface{}) bool {
+			value.(chan *net.MaxFloatUpdate) <- &net.MaxFloatUpdate{
+				Sender:   sender.Bytes(),
+				MaxFloat: maxFloat.Bytes(),
+			}
+			return true
+		},
+	)
+}
+
+func (r *redeemer) MonitorMaxFloat(req *empty.Empty, stream net.TicketRedeemer_MonitorMaxFloatServer) error {
+	// The client address will serve as the ID for the stream
+	p, ok := peer.FromContext(stream.Context())
+	if !ok {
+		return status.Error(codes.Internal, "context is nil")
+	}
+
+	// Make a channel to receive max float updates
+	//  This check allows to overwrite the channel for testing purposes
+	var maxFloatUpdates chan *net.MaxFloatUpdate
+	maxFloatUpdatesI, ok := r.subs.Load(p.Addr.String())
+	if !ok {
+		maxFloatUpdates = make(chan *net.MaxFloatUpdate)
+		r.subs.Store(p.Addr.String(), maxFloatUpdates)
+		glog.Infof("new MonitorMaxFloat subscriber: %v", p.Addr.String())
+	}
+	maxFloatUpdates, ok = maxFloatUpdatesI.(chan *net.MaxFloatUpdate)
+	if !ok {
+		return status.Error(codes.Internal, "maxFloatUpdates is of the wrong type")
+	}
+
+	// Block so that the stream is over a long-lived connection
+	for {
+		select {
+		case maxFloatUpdate := <-maxFloatUpdates:
+			if err := stream.Send(maxFloatUpdate); err != nil {
+				if err == io.EOF {
+					r.subs.Delete(p.Addr.String())
+					return status.Error(codes.Internal, err.Error())
+				}
+				glog.Errorf("Unable to send maxFloat update to client=%v err=%v", p.Addr.String(), err)
+			}
+		case <-r.quit:
+			return nil
+		case <-stream.Context().Done():
+			return nil
+		}
+	}
+}
+
+type redeemerClient struct {
+	rpc      net.TicketRedeemerClient
+	maxFloat map[ethcommon.Address]*big.Int
```

Comment:
Run a watchdog to clean up this map? How to init this value ? Add a remote call ?

Reply:
We should clean up this map similar to what we do in the current `SenderMonitor` implementation and call `Clear()` on the `pm.SenderManager` instance so that cached sender info can be cleared.

Reply:
added a cleanup loop in c5609320614c7c842151f2a41b395e1725c76aa5